File: python/1-9/problem4.py
def find_max_palindrome(minimum=100, maximum=999):
    max_palindrome = 0
    for i in range(maximum, minimum, -1):
        for r in range(maximum, minimum, -1):
            x = str(i * r)
            if x == x[::-1]:
                if int(x) > max_palindrome:
                    max_palindrome = int(x)

    return max_palindrome

print(find_max_palindrome())

File: ruby/problem4.rb
# A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

# Find the largest palindrome made from the product of two 3-digit numbers.

def max_palindrome(minimum: 100, maximum: 999)
  (minimum...maximum).reverse_each.filter_map do |left|
    (minimum...maximum).reverse_each.filter_map do |right|
      x = (left * right).to_s
      x.to_i if x == x.reverse
    end.max
  end.max
end

puts max_palindrome
